File: android/app/src/main/java/com/vergipro/mobile/core/session/SessionController.kt
package com.vergipro.mobile.core.session

import com.vergipro.mobile.core.security.CredentialStore

enum class SessionState { SignedOut, Restoring, Active, Locked }

class SessionController(private val credentialStore: CredentialStore) {
    @Volatile private var accessToken: String? = null
    @Volatile var state: SessionState = SessionState.SignedOut
        private set

    fun currentAccessToken(): String? = accessToken

    @Synchronized
    fun establish(newAccessToken: String, refreshToken: String) {
        credentialStore.saveRefreshToken(refreshToken.encodeToByteArray())
        accessToken = newAccessToken
        state = SessionState.Active
    }

    @Synchronized
    fun beginRestore(): ByteArray? {
        state = SessionState.Restoring
        return credentialStore.readRefreshToken()
    }

    @Synchronized
    fun completeRestore(newAccessToken: String, rotatedRefreshToken: String) {
        establish(newAccessToken, rotatedRefreshToken)
    }

    @Synchronized
    fun lock() {
        accessToken = null
        state = SessionState.Locked
    }

    @Synchronized
    fun signOut() {
        accessToken = null
        credentialStore.deleteRefreshToken()
        state = SessionState.SignedOut
    }
}

File: ios/VergiPro/Core/Session/SessionController.swift
import Foundation

actor SessionController {
    static let shared = SessionController(credentialStore: KeychainCredentialStore())
    enum State: Equatable { case signedOut, restoring, active, locked }

    private(set) var state: State = .signedOut
    private var accessToken: String?
    private let credentialStore: CredentialStore
    private let accountKey = "primary-device-session"

    init(credentialStore: CredentialStore) {
        self.credentialStore = credentialStore
    }

    func currentAccessToken() -> String? { accessToken }

    func establish(accessToken: String, refreshToken: String) throws {
        try credentialStore.saveRefreshToken(Data(refreshToken.utf8), account: accountKey)
        self.accessToken = accessToken
        state = .active
    }

    func beginRestore() throws -> Data? {
        state = .restoring
        return try credentialStore.readRefreshToken(account: accountKey)
    }

    func completeRestore(accessToken: String, rotatedRefreshToken: String) throws {
        try establish(accessToken: accessToken, refreshToken: rotatedRefreshToken)
    }

    func lock() {
        accessToken = nil
        state = .locked
    }

    func signOut() throws {
        accessToken = nil
        try credentialStore.deleteRefreshToken(account: accountKey)
        state = .signedOut
    }
}
